Report/sections/4_testing/unit_testing.tex

```diff
@@ -0,0 +1,106 @@
+\section{Unit Testing}
+In implementing unit testing, we have followed a few approaches.
+In general, we have attempted to follow the Act-Arrange-Assert (AAA)
+methodology, as well as the naming convention 
+\texttt{MethodName\_Scenario\_ExpectedResult}.
+
+We did not use Test-Driven Development, instead resorting to testing
+when we've finished the implementation of each part. While not ideal, it
+contributed to us catching bugs and fixing errors that we wouldn't
+otherwise have caught.
+
+To illustrate our approach to testing, we will draw examples from our
+AST Builder tests.
+
+Our approach is to traverse the AST through a visitor. First,
+we had the test fixture class inherit from an AST visitor interface.
+Then we implemented traversal between nodes, where each node knows how
+to visit the node(s) below.
+
+To test the visitor, we wrote some test code in \dazel{}, which can be seen in \snipref{lst:asttestcode}. We made sure that
+the test fixture class could load this test code and generate a parse
+tree from it.
+
+In each \texttt{Visit} method, we call \texttt{Assert} which tests
+that the node contains the correct information in relation to the
+aforementioned \dazel{} test code.
+
+For example, the visit to a \texttt{GameObject} checks that it has the
+right type (line \ref{code:stcheck}), identifier (line \ref{code:sicheck}), and contents (line \ref{code:sccheck}). Then it goes on to check that
+\texttt{GameObjectContent} are of the right types, and that they have
+correct amount of statements in their scopes. Lastly, it starts visiting
+those such that a similar process can happen to the remaining nodes (line \ref{code:sprocc}).
+
+\begin{lstlisting}[caption={AST Builder \dazel{} test code}, label={lst:asttestcode}]
+    "Screen SampleScreen1" + // Test GameObject
+    "{" +
+    "   Map" + // Test GameObjectContent & GameObjectContentType
+    "   {" +
+    "       Size(30, 24);" + // Test function invocation
+    "       SomeVar1 = 2.0;" + // Test assignment & floats
+    "       { " + // Test statement block
+    "           SomeVar2 = 3 + 3 / 3;" + // Test expressions & integers
+    "           x = SomeVar2;" + // Test identifier values
+    "       }" +
+    "       let = SampleScreen1.Exits.exit1;" + // Test member access 
+    "       arr = [1, 2, 3];" + // Test arrays
+    "   }" +
+    "" +
+    "   Entities" +
+    "   {" +
+    "       SpawnEntity(Skeleton1, [4, 5]);" +
+    "   }" +
+    "}";
+    \end{lstlisting}
+
+
+\begin{lstlisting}[caption={GameObject visitor code for testing the AST Builder}, label={lst:gameobjecttest},escapechar=|, language=CSharp]
```

Comment:
We need to update this code once the UnityImplementation branch is in because these are the old names for classes.

Reply:
Added a task to our board.